### PR TITLE
Add mixtral command

### PR DESCRIPTION
**Pull request by glcraft**
Add mixtral command

The command makes a Replicate prediction over Mixstral 8x7b LLM model.

**NOTE** : Replicate is a paid service, which means it's not free (at least forever). But that said, it's pretty cheap and it's a good service. [See Pricing at the bottom of this page](https://replicate.com/mistralai/mixtral-8x7b-instruct-v0.1)

### CODE REVIEW

File: src/commands/ai/mixtral.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+export const MIXTRAL: Command = {
+    data: new SlashCommandBuilder()
+        .setName('mixtral')
+        .setDescription('Génère du texte avec Mixtral 8x7b!')
+        .addStringOption((option) => option.setName('prompt').setDescription('Le prompt').setRequired(true))
+        .addNumberOption((option) =>
+            option
+                .setName('max_new_tokens')
+                .setDescription('Le nombre maximum de jetons que le modèle doit générer en sortie (1024 par défaut)')
```
Comment (Arthur-Devs) [inline src/commands/ai/mixtral.ts]: Utiliser valeur max et valeur minimum plutôt que de tester après l'envoi de la commande

**Review by Arthur-Devs — CHANGES_REQUESTED** [review]

File: src/commands/ai/replicate.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { ENV } from '@env';
```
Comment (vic1707) [inline src/commands/ai/replicate.ts]: file in a "globals" dir or something ? :thinking:

File: src/commands/ai/mixtral.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    max_new_tokens: 1024,
+    temperature: 0.6,
+    top_p: 0.9,
+    top_k: 50,
+    presence_penalty: 0,
+    frequency_penalty: 0
+    prompt_template: "<s>[INST] {prompt} [/INST] ",
+};
+*/
+
```
Comment (vic1707) [inline src/commands/ai/mixtral.ts]: rm

File: src/commands/ai/mixtral.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                .setDescription('Pénalité de fréquence (0 par défaut)')
+                .setMinValue(-2)
+                .setMaxValue(2)
+                .setRequired(false),
+        ),
+    async execute(interaction) {
+        await interaction.deferReply();
+        const replyError = async (msgError: string): Promise<void> => {
+            await interaction.editReply(msgError);
+        };
```
Comment (vic1707) [inline src/commands/ai/mixtral.ts]: global function/utility class ?

File: src/commands/ai/mixtral.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        const input = {
+            prompt: interaction.options.get('prompt')?.value as string,
+            max_new_tokens: (interaction.options.get('max_new_tokens')?.value as number) || 1024,
+            temperature: (interaction.options.get('temperature')?.value as number) || 0.6,
+            top_p: (interaction.options.get('top_p')?.value as number) || 0.9,
+            top_k: (interaction.options.get('top_k')?.value as number) || 50,
+            presence_penalty: (interaction.options.get('presence_penalty')?.value as number) || 0,
+            frequency_penalty: (interaction.options.get('frequency_penalty')?.value as number) || 0,
+            prompt_template: '<s>[INST] {prompt} [/INST] ',
+        };
```
Comment (vic1707) [inline src/commands/ai/mixtral.ts]: better have a global object on the top of the file with default options then 
```ts
const cmdInput = { /*...*/ };
const finalInput = { ...DEFAULT_CFG, ...cmdInput};
```

File: src/commands/ai/mixtral.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        let msg = '';
+        let last_time = Date.now();
+        for await (const event of replicate.stream('mistralai/mixtral-8x7b-instruct-v0.1', { input })) {
+            if (event.event === 'output') {
+                msg += event.data;
+            }
```
Comment (vic1707) [inline src/commands/ai/mixtral.ts]: maybe a switch to be more exhaustive ?

File: src/commands/ai/mixtral.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+            if (Date.now() - last_time > 500 && msg !== '') {
+                await interaction.editReply(msg);
+                last_time = Date.now();
+            }
```
Comment (vic1707) [inline src/commands/ai/mixtral.ts]: can do better than 2 ifs in a row ? are they mergeable ?

**Review by vic1707 — CHANGES_REQUESTED** [review]

File: src/commands/ai/mixtral.ts (hunk 7 above)
Comment (glcraft) [reply src/commands/ai/mixtral.ts]: nope they are two distinguish statements

File: src/commands/ai/mixtral.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+export const MIXTRAL: Command = {
+    data: new SlashCommandBuilder()
+        .setName('mixtral')
+        .setDescription('Génère du texte avec Mixtral 8x7b!')
+        .addStringOption((option) => option.setName('prompt').setDescription('Le prompt').setRequired(true))
+        .addNumberOption((option) =>
+            option
+                .setName('max_new_tokens')
+                .setDescription('Le nombre maximum de jetons que le modèle doit générer en sortie (1024 par défaut)')
```
Comment (vic1707) [inline src/commands/ai/mixtral.ts]: the other command (ping) is in english, do we do 100% english, french, localized ?

File: src/commands/ai/tellme.ts [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                .setDescription('Le nombre maximum de jetons que le modèle doit générer en sortie (1024 par défaut)')
+                .setMinValue(0)
+                .setMaxValue(10000)
+                .setRequired(false),
+        ),
+    execute: async (interaction) => {
+        await interaction.deferReply();
+        const input = {
+            prompt: interaction.options.get('prompt')?.value as string,
+            max_new_tokens: (interaction.options.get('max_new_tokens')?.value as number) || 1024,
```
Comment (vic1707) [inline src/commands/ai/tellme.ts]: 1024, magic number

File: src/commands/ai/tellme.ts [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        let last_time = Date.now();
+        for await (const event of replicate.stream('mistralai/mixtral-8x7b-instruct-v0.1', { input })) {
+            if (event.event === 'output') {
+                msg += event.data;
+            }
+            if (Date.now() - last_time > 500 && msg !== '') {
+                await interaction.editReply(msg);
+                last_time = Date.now();
+            }
+        }
```
Comment (vic1707) [inline src/commands/ai/tellme.ts]: you use that piece of code in two files, you probably shoud move it to the `replicate.ts` or a utils file

File: src/commands/ai/mixtral.ts (hunk 8 above)
Comment (glcraft) [reply src/commands/ai/mixtral.ts]: The server we target for is in french.
So either we do both with localization feature, either all in french. The french language is **mandatory**.

File: src/commands/ai/mixtral.ts (hunk 6 above)
Comment (glcraft) [reply src/commands/ai/mixtral.ts]: The whole options are either [`output`, `logs`, `error` or `done`](https://github.com/replicate/replicate-javascript#server-sent-events).
I did not care about the other options.
So what do you think ?

File: src/commands/ai/mixtral.ts (hunk 8 above)
Comment (vic1707) [reply src/commands/ai/mixtral.ts]: k will do another PR to verify all of this

**Review by vic1707 — DISMISSED** [review]

**Review by Arthur-Devs — DISMISSED** [review]

File: .env.example [hunk 11]
```diff
@@ -1,3 +1,4 @@
 TOKEN=''
 CLIENT_ID=''
 SERVER_ID=''
+REPLICATE_TOKEN=''
```
Comment (vic1707) [inline .env.example:4]: can you make the updates to the Deploy CI/CD file ?

File: .vscode/launch.json [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            "type": "bun",
+            "internalConsoleOptions": "neverOpen",
+            "request": "attach",
+            "name": "Attach Bun",
+            "url": "ws://localhost:6499/",
+            "stopOnEntry": false
+        }
+    ]
+}
```
Comment (vic1707) [inline .vscode/launch.json]: ok to merge if you're sure that all of this works

File: src/bot.ts [hunk 13]
```diff
@@ -100,7 +100,6 @@ export class Bot extends Client {
      */
     private setUpEvents(): void {
         for (const event of EVENTS) {
-            console.log(`Setting up event: ${event.name}`);
             // else TSserver will complain about the type of event.execute
```
Comment (vic1707) [inline src/bot.ts:103]: haha, was ok to add it, now you remove xD

File: src/utils/replicate/index.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import { ENV } from '@env';
+import Replicate from 'replicate';
+
+export const replicate = new Replicate({
```
Comment (vic1707) [inline src/utils/replicate/index.ts]: may i ask to capitalize ? 

the `import { replicate } from '.';`
looks weird when there is `import Replicate from 'replicate';` in the same project x)

**Review by vic1707 — DISMISSED** [review]
really small things but non blocking

File: src/bot.ts (hunk 13 above)
Comment (glcraft) [reply src/bot.ts:103]: weird, it has already been removed 🤔

**Review by vic1707 — DISMISSED** [review]

**Review by vic1707 — APPROVED** [review]

**Pull request merged**